File: binarytreelongestconsecutivesequence.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int longestConsecutive(TreeNode* root) {
        if (!root) return 0;
        return max(dfs(root->left, root, 1), dfs(root->right, root, 1));
    }
    int dfs(TreeNode* root, TreeNode* pre, int depth) {
        if (!root) return depth;
        if (pre->val+1 == root->val) depth += 1;
        else depth = 1;
        pre = root;
        return max(depth, max(dfs(root->left, pre, depth), dfs(root->right, pre, depth)));
    }

};

File: python/binarytreelongestconsecutivesequence.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def longestConsecutive(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        return max(self.dfs(root.left, root.val, 1, 1), self.dfs(root.right, root.val, 1, 1))
        
    def dfs(self, root, pre, maximum, path):
        if not root:
            return 1
        if pre+1 == root.val:
            path += 1
            maximum = max(maximum, path)
        else:
            path = 1
        leftmax = self.dfs(root.left, root.val, maximum, path)
        rightmax = self.dfs(root.right, root.val, maximum, path)
        return max(maximum, leftmax, rightmax)
